doc/lftp.1
.\"
.\" lftp.1 - Sophisticated file transfer program
.\"
.\" This file is part of lftp.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\"-------
.\" Sp  space down the interparagraph distance
.\"-------
.de Sp
.sp \\n(Ddu
..
.\"-------
.\" Ds  begin a display, indented .5 inches from the surrounding text.
.\"
.\" Note that uses of Ds and De may NOT be nested.
.\"-------
.de Ds
.Sp
.RS
.nf
..
.\"-------
.\" De  end a display (no trailing vertical spacing)
.\"-------
.de De
.RE
.fi
..
.TH lftp 1 "11 June 2022"
.SH NAME
lftp \- Sophisticated file transfer program
.SH SYNTAX
.B lftp
[\fB\-d\fR] [\fB\-e \fIcmd\fR] [\fB\-p \fIport\fR]
[\fB\-u \fIuser\fR[\fB,\fIpass\fR]] [\fIsite\fR]
.br
.BI "lftp \-f " script_file
.br
.BI "lftp \-c " commands
.br
.B lftp \-\-version
.br
.B lftp \-\-help

.SH VERSION
This man page documents lftp version 4.8.1.

.SH "DESCRIPTION"
\fBlftp\fR is a file transfer program that allows sophisticated FTP, HTTP
and other connections to other hosts. If \fIsite\fR is specified then lftp
will connect to that site otherwise a connection has to be established with
the open command.
.PP
\fBlftp\fP can handle several file access methods - FTP, FTPS, HTTP, HTTPS,
HFTP, FISH, SFTP and file (HTTPS and FTPS are only available when lftp is
compiled with GNU TLS or OpenSSL library). You can specify the method to use in `open
URL' command, e.g. `open http://www.us.kernel.org/pub/linux'. HFTP is
ftp-over-http-proxy protocol. It can be used automatically instead of FTP
if ftp:proxy is set to `http://proxy[:port]'. Fish is a protocol working
over an ssh connection to a unix account. SFtp is a protocol implemented
in SSH2 as SFTP subsystem.
.PP
Besides FTP-like protocols, lftp has support for BitTorrent protocol
as `torrent' command. Seeding is also supported.

.PP
Every operation in \fBlftp\fP is reliable, that is any non-fatal error is
handled properly and the operation is repeated. So if downloading breaks, it
will be restarted from the point automatically. Even if FTP server
does not support the REST command, \fBlftp\fP will try to retrieve the file from
the very beginning until the file is transferred completely.

\fBlftp\fP has shell-like command syntax allowing you to launch several
commands in parallel in background (&). It is also possible to group
commands within () and execute them in background. All background jobs
are executed in the same single process. You can bring a foreground
job to background with ^Z (c-z) and back with command `wait' (or `fg' which
is alias to `wait'). To list running jobs, use command `jobs'. Some
commands allow redirecting their output (cat, ls, ...) to file or via
pipe to external command. Commands can be executed conditionally based
on termination status of previous command (&&, ||).
.PP
If you exit \fBlftp\fP before all jobs are not finished yet, \fBlftp\fP will move
itself to nohup mode in background. The same thing happens with a
real modem hangup or when you close an xterm.
.PP
\fBlftp\fP has built-in mirror which can download or update a whole directory
tree. There is also reverse mirror (mirror \-R) which uploads or
updates a directory tree on server. Mirror can also synchronize directories
between two remote servers, using FXP if available.
.PP
There is command `at' to launch a job at specified time in current
context, command `queue' to queue commands for sequential execution
for current server, and much more.
.PP
On startup, \fBlftp\fP executes \fI/etc/lftp.conf\fP and then \fI~/.lftprc\fP and
\fI~/.lftp/rc\fP (or \fI~/.config/lftp/rc\fP if \fI~/.lftp\fP does not exist).
You can place aliases and `set' commands there. Some
people prefer to see full protocol debug, use `debug' to turn the
debug on. Use `debug 3' to see only greeting messages and error
messages.
.PP
\fBlftp\fP has a number of settable variables. You can use `set \-a' to see
all variables and their values or `set \-d' to see list of defaults.
Variable names can be abbreviated and prefix can be omitted unless the
rest becomes ambiguous.
.PP
If lftp was compiled with OpenSSL (configure \-\-with\-openssl) it includes software developed
by the OpenSSL Project for use in the OpenSSL Toolkit. (http://www.openssl.org/)

.SS Commands
.PP

.BI ! " shell command"
.PP
Launch shell or shell command.
.PP
.Ds
!ls
.De
.PP
To do a directory listing of the local host.

.B alias
.RI " [" name " [" value ]]
.PP
Define or undefine alias \fIname\fP. If \fIvalue\fP is omitted, the alias is
undefined, else it takes the value \fIvalue\fP. If no argument is given
the current aliases are listed.
.PP
.Ds
alias dir ls \-lF
alias less zmore
.De

.BR at " \fItime\fP [ \-\- \fIcommand\fP ] "
.PP
Wait until the given time and execute given (optional) command. See also \fBat\fR(1).

.B attach
.RI " [" PID ]
.PP
Attach the terminal to specified backgrounded lftp process.

.B bookmark
.RI " [" subcommand ]
.PP
The bookmark command controls bookmarks.

Site names can be used in the \fIopen\fP command directly as-is or in
any command that accepts input URLs using the \fIbm:site/path\fP format.
.Sp
.RS
.TS
l	lx	.
add <name> [<loc>]	T{
add current place or given location to bookmarks and bind to given name
T}
del <name>	remove bookmark with name
edit	start editor on bookmarks file
import <type>	import foreign bookmarks
list	list bookmarks (default)
.TE
.RE
.P
.B cache
.RI " [" subcommand ]
.PP
The cache command controls local memory cache.
The following subcommands are recognized:
.Sp
.RE
.TS
l	lx	.
stat	print cache status (default)
on|off	turn on/off caching
flush	flush cache
size \fIlim\fP	set memory limit, -1 means unlimited
expire \fINx\fP	T{
set cache expiration time to \fIN\fP seconds (\fIx\fP=s) minutes (\fIx\fP=m) hours (\fIx\fP=h) or days (\fIx\fP=d)
T}
.TE
.RE
.P
.BR cat " \fIfiles\fP"
.PP
cat outputs the remote file(s) to stdout.  (See also \fBmore\fR,
\fBzcat\fR and \fBzmore\fR)

.BR cd " rdir"
.PP
Change current remote directory.  The previous remote directory is
stored as `\-'. You can do `cd \-' to change the directory back.
The previous directory for each site is also stored on disk,
so you can do `open site; cd \-' even after lftp restart.

.BR chmod " [" OPTS "] " "\fImode files...\fP"
.PP
Change permission mask on remote files. The mode can be an octal number or a symbolic mode (see chmod(1)).
.PP
.Sp
.RS
.TS
l1	l	lx	.
-c,	--changes	like verbose but report only when a change is made
-f,	--quiet	suppress most error messages
-v,	--verbose	output a diagnostic for every file processed
-R,	--recursive	change files and directories recursively
.TE
.RE
.P
.BR close " [" \-a "]"
.PP
Close idle connections.  By default only with the current server, use
\-a to close all idle connections.

.BR cls " [" OPTS "] " \fIfiles...\fP
.PP
`cls' tries to retrieve information about specified files or directories
and outputs the information according to format options. The difference between
`ls' and `cls' is that `ls' requests the server to format file listing, and
`cls' formats it itself, after retrieving all the needed information.
.PP
.Sp
.RS
.TS
l1	l	lx	.
\-1		single-column output
\-a,	\-\-all	show dot files
\-B,	\-\-basename	show basename of files only
	\-\-block\-size=SIZ	use SIZ-byte blocks
\-d,	\-\-directory	T{
list directory entries instead of contents
T}
\-F,	\-\-classify	T{
append indicator (one of /@) to entries
T}
\-h,	\-\-human\-readable	T{
print sizes in human readable format (e.g., 1K)
T}
	\-\-si	T{
likewise, but use powers of 1000 not 1024
T}
\-k,	\-\-kilobytes	T{
like \-\-block\-size=1024
T}
\-l,	\-\-long	T{
use a long listing format
T}
\-q,	\-\-quiet	don't show status
\-s,	\-\-size	print size of each file
	\-\-filesize	T{
if printing size, only print size for files
T}
\-i,	\-\-nocase	T{
case-insensitive pattern matching
T}
\-I,	\-\-sortnocase	T{
sort names case-insensitively
T}
\-D,	\-\-dirsfirst	T{
list directories first
T}
	\-\-sort=OPT	T{
"name", "size", "date"
T}
\-S		sort by file size
	\-\-user, \-\-group,
	\-\-perms, \-\-date,
	\-\-linkcount, \-\-links	show individual fields
	\-\-time\-style=STYLE	T{
use specified time format
T}
.TE
.RE
.P
.BR command " \fIcmd args...\fP
.PP
execute given command ignoring aliases.

.BR debug " [" OPTS "] " \fIlevel\fP | off
.PP
Switch debugging to \fIlevel\fP or turn it off. Options:
.Sp
.RS
.TS
l1	l	lx	.
\-T	truncate output file
\-o <file>	redirect debug output to the file
\-c	show message context
\-p	show PID
\-t	show timestamps
.TE
.RE
.P
.BR du " [" OPTS "] " \fIpath...\fP
.PP
Summarize disk usage. Options:
.Sp
.RS
.TS
l1	l	lx	.
\-a,	\-\-all	T{
write counts for all files, not just directories
T}
	\-\-block\-size=SIZ	use SIZ-byte blocks
\-b,	\-\-bytes	print size in bytes
\-c,	\-\-total	produce a grand total
\-d,	\-\-max\-depth=N	T{
print the total for a directory (or file, with \-\-all) only if it is N or
fewer levels below the command line argument;  \-\-max\-depth=0 is the same as
\-\-summarize
T}
\-F,	\-\-files	print number of files instead of sizes
\-h,	\-\-human\-readable	T{
print sizes in human readable format (e.g., 1K 234M 2G)
T}
\-H,	\-\-si	likewise, but use powers of 1000 not 1024
\-k,	\-\-kilobytes	like \-\-block\-size=1024
\-m,	\-\-megabytes	like \-\-block\-size=1048576
\-S,	\-\-separate\-dirs	do not include size of subdirectories
\-s,	\-\-summarize	display only a total for each argument
	\-\-exclude=PAT	exclude files that match PAT
.TE
.RE
.P
.BR echo " [" \-n "] \fIstring\fR"
.PP
Prints (echos) the given string to the display.

.BR edit " [" OPTS "] " \fIfile\fP
.PP
Retrieve remote file to a temporary location, run a local editor on it
and upload the file back if changed. Options:
.Sp
.RS
.TS
l1	l	lx	.
\-k	keep the temporary file
\-o <temp>	explicit temporary file location
.TE
.RE
.P
.BR eval " [" -f " \fIformat\fR ] " \fIargs...\fR
.PP
without -f it executes given arguments as a command. With -f, arguments
are transformed into a new command. The format can contain plain text and
placeholders $0...$9 and $@, corresponding to the arguments.

.BR exit " [" bg "]"
.RB [ top ]
.RB [ parent ]
.RB [ kill ]
.RI [ code ]
.PP
exit will exit from lftp or move to background if there are active jobs. If
no job is active, \fIcode\fP is passed to operating system as lftp's
termination status. If \fIcode\fP is omitted, the exit code of last
command is used.
.PP
`exit bg' forces moving to background when cmd:move-background is false.
`exit top' makes top level `shell' (internal lftp command executor) terminate.
`exit parent' terminates the parent shell when running a nested script.
`exit kill' kills all numbered jobs before exiting. The options can be combined, e.g.
`at 08:00 \-\- exit top kill &' kills all jobs and makes lftp exit at specified time.

.B fg
.PP
Alias for `wait'.

.BR find " [" OPTS "] " \fIdirectory...\fP
.PP
List files in the directory (current directory by default) recursively.
This can help with servers lacking ls \-R support. You can redirect output
of this command. Options:
.Sp
.RS
.TS
l1	l	lx	.
\-d \fIMD\fP,	\-\-max\-depth=\fIMD\fP	specify maximum scan depth

NEWS
Version 4.9.3 - 2024-11-08

* fixed sftp out-of-order problem.
* fixed udp tracker looping.
* fixed a few coredumps.
* fixed compilation in some cases.
* translations updated.

Version 4.9.2 - 2020-08-13

* mkdir: fixed exit code with -f option.
* ftp: made ftp:use-pret setting tri-boolean.
* get/mget/put/mput: don't try next files after error if cmd:fail-exit is true.
* get/mget: fixed -O option with remote URL and xfer:use-temp-file being true.
* mirror: disallow empty patterns; don't delete "..".
* mirror: fixed --on-change with --reverse.
* sftp: fixed a bug with truncated files when packets are reordered (finally).

Version 4.9.1 - 2020-01-15

* get1: fixed creating the target directory.
* get1: fixed renaming the temp file.
* fixed a coredump on startup when IPv6 is not supported or configured.
* fixed build with --disable-ipv6.
* fixed build with Sun Studio.

Version 4.9.0 - 2019-12-21

* ftp: add support for CEPR command.
* ftp: enabled TLS by default.
* ftp: fixed OPTS MLST format.
* ftps: apply rfc1738 rules.
* ftp: stop sending ALLO commands by default.
* http: fixed redirection to absolute path.
* http: improved index parsing performance.
* http: revert to original URL when temporary redirects led to an error.
* mirror --no-empty-dirs: exclude empty dirs so that chmod is not done on them.
* net: disable inet6 if there is no public ipv6 address.
* net: lowered reconnect intervals.
* sftp, fish: fixed yes/no/[fingerprint] and passphrase recognition.
* sftp: fixed a bug with truncated files when packets are reordered.
* torrent: fixed a coredump when all dht ports are busy.
* fixed file list sorting when some file names start with tilde.
* fixed retry counting when a site has multiple addresses.
* fixed "version" command for certain openssl versions.

Version 4.8.4 - 2018-08-01

* fixed a security vulnerability with "file:" file names.
* fixed mirror --flat.
* http: extract links from <source> tags.
* fixed upload of zero-length files over ftps.
* fixed assert on "mput -d".
* fixed a core dump.
* avoid multiple backup files of DHT cache.
* translations updated (uk, zh_CN).

Version 4.8.3 - 2017-10-09

* fixed compilation with older C++ compilers.
* fixed 2 other core dumps on exit (incorrect order of deallocation).

Version 4.8.2 - 2017-09-25

* fixed compilation on FreeBSD and OS X.
* fixed a memory leak.
* fixed a core dump on exit.

Version 4.8.1 - 2017-09-12

* improved rate limiting to allow per-host limits.
* ftp: make prefer-epsv default "no".
* ftp: fixed quote command and ftp:use-stat-for-list setting.
* switched to libidn2.
* fixed build with LibreSSL.
* fixed configure --disable-rpath.
* fixed coredump when a parent directory was deleted.

Version 4.8.0 - 2017-07-10

* mirror: improved performance of --scan-all-first for big trees.
* mirror: new --flat option to flatten the target directory structure.
* mmv: new command for file moving; redirect mv to mmv in certain cases.
* fixed compilation with newer openssl (1.1.0 and later).
* du: allow multiple --exclude options to be combined.
* new setting cmd:nullglob for `glob' command prefix.
* http: use proppatch to set last-modified property.
* new settings net:connection-limit-timer and ftp:too-many-re.
* ftp: dynamically ajust connection limit.
* ftp: fixed core dump on LINK/SYMLINK when the command is not supported.
* get1: fixed -o option.
* sftp,fish: connect-program setting is now passed to the shell for execution.
* get/mget/put/mput: add -P option for parallel transfers and long options.
* appimage: new make target for making an AppImage file.
* fixed "local glob".

Version 4.7.7 - 2017-03-07

* mirror: fixed coredump when source directory does not exist.
* mirror: don't create target directory if can't enter to the source directory.
* ftp: fixed a rare hang when a NOOP was sent between "transfer ok" reply
  and EOF on data socket.
* fixed xfer:log setting (compatibility alias).
* ftp: don't use EPSV with a proxy.

Version 4.7.6 - 2017-02-15

* obsolete settings xfer:log, xfer:log-file, xfer:max-log-size are now
  aliased to log:enabled/xfer, log:file/xfer and log:max-size/xfer.
* fixed a missing SFTP status message.
* fixed a coredump when opening "slot:name" without a path in the slot.
* fixed XDG directories description in the man page.
* fixed off-by-one month error in apache file listing with ISO dates.
* fixed compilation on some systems.
* updated zh_TW translation.

Version 4.7.5 - 2017-01-01

* http: recognize apache listing with ISO date/time.
* ipv6: ignore link-local address without a scope.
* new settings log:prefix-{recv,send,note,error}.
* documentation and help improvements.
* mirror: fixed recursion mode initialization.
* mirror: fixed assertion failure when creating remote symlinks.
* mirror: show chmod errors when verbose.
* fixed debug command to switch output to stderr if no output file given.
* ftp: add "site mkdir" support for "mkdir -p".
* ftp: fixed a long delay in "mode z" with ProFTPD.
* ftp: fixed a failed assertion in "mode z".

Version 4.7.4 - 2016-11-16

* mirror: new option --directory (-F) for source directories glob pattern.
* mirror: fixed coredump with --depth-first + --recursion=newer.
* mirror: create inaccessible directories anyway.
* pget, torrent: avoid long delays in posix_fallocate(3) call.
* new setting dns:name for host aliases.
* ftp: allow SSL after certain types of ftp proxy.
* http: fixed interpretation of links without a protocol.
* cls: fixed glob metacharacter unquoting.
* sftp: fixed mirroring of files with a leading tilde.
* fixed transfer logging.

Version 4.7.3 - 2016-07-15

* mirror: don't add source directory base name to the target if the source
  ends with a slash.
* mirror: fixed transfer count tracking in --scan-all-first mode.

src/lftp.cc
/*
 * lftp - file transfer program
 *
 * Copyright (c) 1996-2013 by Alexander V. Lukyanov (lav@yars.free.net)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include "modconfig.h"

#include "trio.h"
#include <sys/types.h>
#include <sys/stat.h> // for mkdir()
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <locale.h>
#include <ctype.h>

#include "xstring.h"
#include "xmalloc.h"
#include "alias.h"
#include "CmdExec.h"
#include "SignalHook.h"
#include "GetPass.h"
#include "History.h"
#include "log.h"
#include "DummyProto.h"
#include "ResMgr.h"
#include "LsCache.h"
#include "IdNameCache.h"
#include "LocalDir.h"
#include "ConnectionSlot.h"
#include "misc.h"
#include "ArgV.h"
#include "attach.h"

#include "configmake.h"

#include "lftp_rl.h"
#include "complete.h"

CDECL_BEGIN
#include <glob.h>
CDECL_END

#define top_exec CmdExec::top

void  hook_signals()
{
   SignalHook::DoCount(SIGHUP);
   SignalHook::Ignore(SIGTTOU);
   ProcWait::Signal(true);
}

ResDecl res_save_cwd_history
   ("cmd:save-cwd-history","yes",ResMgr::BoolValidate,ResMgr::NoClosure);
ResDecl res_save_rl_history
   ("cmd:save-rl-history","yes",ResMgr::BoolValidate,ResMgr::NoClosure);
ResDecl res_stifle_rl_history
   ("cmd:stifle-rl-history","500",ResMgr::UNumberValidate,ResMgr::NoClosure);

class ReadlineFeeder : public CmdFeeder, private ResClient
{
   bool tty:1;
   bool ctty:1;
   bool add_newline:1;
   int eof_count;
   xstring_c cmd_buf;
   xstring_c for_history;

   static bool readline_inited;
   void readline_init()
   {
      if(readline_inited)
	 return;
      readline_inited=true;
      lftp_readline_init();
      lftp_rl_read_history();
      if(for_history)
      {
	 lftp_add_history_nodups(for_history);
	 for_history.set(0);
      }
      Reconfig(0);
   }

public:
   ReadlineFeeder(const ArgV *args)
   {
      tty=isatty(0);
      ctty=(tcgetpgrp(0)!=(pid_t)-1);
      add_newline=false;
      eof_count=0;
      if(args && args->count()>1)
	 args->CombineQuotedTo(for_history);
   }
   virtual ~ReadlineFeeder()
   {
      if(readline_inited)
      {
	 if(res_save_cwd_history.QueryBool(0))
	    cwd_history.Save();
	 if(res_save_rl_history.QueryBool(0))
	    lftp_rl_write_history();
      }
   }
   bool IsInteractive() const
   {
      return tty;
   }
   bool RealEOF()
   {
      return !tty || eof_count>3;
   }

   const char *NextCmd(class CmdExec *exec,const char *prompt)
   {
      if(add_newline)
      {
	 add_newline=false;
	 return "\n";
      }

      ::completion_shell=exec;
      ::remote_completion=exec->remote_completion;

      if(tty)
      {
	 readline_init();

	 if(ctty) // controlling terminal
	 {
	    if(!in_foreground_pgrp())
	    {
	       // looks like we are in background. Can't read from tty
	       exec->Timeout(500);
	       return "";
	    }
	 }

	 SignalHook::ResetCount(SIGINT);
	 cmd_buf.set_allocated(lftp_readline(prompt));
	 xmalloc_register_block(cmd_buf.get_non_const());
	 if(cmd_buf && *cmd_buf)
	 {
	    if(exec->csh_history)
	    {
	       char *history_value0=0;
	       int expanded = lftp_history_expand (cmd_buf, &history_value0);
	       if (expanded)
	       {
		  if(history_value0)
		     xmalloc_register_block(history_value0);
		  xstring_ca history_value(history_value0);

		  if (expanded < 0)
		     fprintf (stderr, "%s\n", history_value.get());

		  /* If there was an error, return nothing. */
		  if (expanded < 0 || expanded == 2)	/* 2 == print only */
		  {
		     exec->Timeout(0);  // and retry immediately
		     return "";
		  }

		  cmd_buf.move_here(history_value);
	       }
	    }
	    lftp_add_history_nodups(cmd_buf);
	 }
	 else if(cmd_buf==0 && exec->interactive)
	    puts("exit");

	 if(cmd_buf==0)
	    eof_count++;
	 else
	    eof_count=0;
      }
      else // not a tty
      {
	 if(exec->interactive)
	 {
	    while(*prompt)
	    {
	       char ch=*prompt++;
	       if(ch!=1 && ch!=2)
		  putchar(ch);
	    }
	    fflush(stdout);
	 }
	 cmd_buf.set_allocated(readline_from_file(0));
      }

      ::completion_shell=0;

      if(cmd_buf && last_char(cmd_buf)!='\n')
      {
	 exec->Timeout(0);
	 add_newline=true;
      }
      return cmd_buf;
   }
   void clear()
      {
	 if(!tty)
	    return;
	 lftp_rl_clear();
      }
   void Reconfig(const char *) {
      lftp_rl_history_stifle(res_stifle_rl_history.Query(0));
   }
};
bool ReadlineFeeder::readline_inited;

#define args	  (parent->args)
#define exit_code (parent->exit_code)
#define output	  (parent->output)
#define session	  (parent->session)
#define eprintf	  parent->eprintf
CMD(history)
{
   enum { READ, WRITE, CLEAR, LIST } mode = LIST;
   const char *fn = NULL;
   static struct option history_options[]=
   {
      {"read",required_argument,0,'r'},
      {"write",required_argument,0,'w'},
      {"clear",no_argument,0,'c'},
      {"list",required_argument,0,'l'},
      {0,0,0,0}
   };

   exit_code=0;
   int opt;
   while((opt=args->getopt_long("+r:w:cl",history_options,0))!=EOF) {
      switch(opt) {
      case 'r':
	 mode = READ;
	 fn = optarg;
	 break;
      case 'w':
	 mode = WRITE;
	 fn = optarg;
	 break;
      case 'c':
	 mode = CLEAR;
	 break;
      case 'l':
	 mode = LIST;
	 break;
      case '?':
	 eprintf(_("Try `help %s' for more information.\n"),args->a0());
	 return 0;
      }
   }

   int cnt = 16;
   if(const char *arg = args->getcurr()) {
      if(!strcasecmp(arg, "all"))
	 cnt = -1;
      else if(isdigit((unsigned char)arg[0]))
	 cnt = atoi(arg);
      else {
	 eprintf(_("%s: %s - not a number\n"), args->a0(), args->getcurr());
	 exit_code=1;
	 return 0;
      }
   }

   switch(mode) {
   case READ:
      if(int err = lftp_history_read(fn)) {
	 eprintf("%s: %s: %s\n", args->a0(), fn, strerror(err));
	 exit_code=1;
      }
      break;

   case WRITE:
      if(int err = lftp_history_write(fn)) {
	 eprintf("%s: %s: %s\n", args->a0(), fn, strerror(err));
	 exit_code=1;
      }
      break;

   case LIST:
      lftp_history_list(cnt);
      break;
   case CLEAR:
      lftp_history_clear();
      break;
   }

   return 0;
}
CMD(attach)
{
   const char *pid_s=args->getarg(1);
   if(!pid_s) {
      xstring& path=AcceptTermFD::get_sock_path(1);
      path.rtrim('1');
      path.append('*');
      glob_t g;
      glob(path, 0, NULL, &g);
      for(size_t i=0; i<g.gl_pathc; i++) {
	 const char *sock_path=g.gl_pathv[i];
	 pid_s=strrchr(sock_path,'-');
	 if(!pid_s)
	    continue;
	 pid_s++;
	 int p=atoi(pid_s);
	 if(p<=1) {
	    pid_s=0;
	    continue;
	 }
	 if(kill(p,0)==-1) {
	    if(errno==ESRCH) {
	       eprintf("%s: removing stale socket `%s'.\n",args->a0(),sock_path);
	       if(unlink(sock_path)==-1)
		  eprintf("%s: unlink(%s): %s\n",args->a0(),sock_path,strerror(errno));
	    }
	    pid_s=0;
	    continue;
	 }
	 pid_s=alloca_strdup(pid_s);
	 break;
      }
      globfree(&g);
      if(!pid_s) {
	 eprintf("%s: no backgrounded lftp processes found.\n",args->a0());
	 return 0;
      }
   }
   int pid=atoi(pid_s);
   SMTaskRef<SendTermFD> term_sender(new SendTermFD(pid));
   while(!term_sender->Done()) {
      SMTask::Schedule();
      SMTask::Block();
   }
   exit_code=0;
   if(term_sender->Failed()) {
      eprintf("%s\n",term_sender->ErrorText());
      exit_code=1;
   }
   return 0;
}
#undef args
#undef exit_code
#undef output
#undef session
#undef eprintf


static void sig_term(int sig)
{
   printf(_("[%u] Terminated by signal %d. %s\n"),(unsigned)getpid(),sig,SMTask::now.IsoDateTime());
   if(top_exec) {
      top_exec->KillAll();
      alarm(30);
      while(Job::NumberOfJobs()>0) {
	 SMTask::Schedule();
	 SMTask::Block();
      }
   }
   exit(1);
}

static void detach()
{
   SignalHook::Ignore(SIGINT);
   SignalHook::Ignore(SIGHUP);
   SignalHook::Ignore(SIGTSTP);

   const char *home=get_lftp_data_dir();
   if(home)
   {
      xstring& log=xstring::get_tmp(home);
      if(access(log,F_OK)==-1)
	 log.append("_log");
      else
	 log.append("/log");

      int fd=open(log,O_WRONLY|O_APPEND|O_CREAT,0600);
      if(fd>=0)
      {
	 dup2(fd,1); // stdout
	 dup2(fd,2); // stderr
	 if(fd!=1 && fd!=2)
	    close(fd);
      }
      const char *c="debug";
      ResMgr::Set("log:show-pid",c,"yes");
      ResMgr::Set("log:show-time",c,"yes");
      ResMgr::Set("log:show-ctx",c,"yes");
   }
   close(0);	  // close stdin.
   open("/dev/null",O_RDONLY); // reopen it, just in case.

#ifdef HAVE_SETSID
   setsid();	  // start a new session.
#endif

   SignalHook::Handle(SIGTERM,sig_term);
}

static void move_to_background()
{
   // notify jobs
   Job::lftpMovesToBackground_ToAll();
   // wait they do something, but no more than 1 sec.
   SMTask::RollAll(TimeInterval(1,0));
   // if all jobs terminated, don't really move to bg.
   if(Job::NumberOfJobs()==0)
      return;

   top_exec->AtBackground();
   top_exec->WaitDone();
   if(Job::NumberOfJobs()==0)
      return;

   fflush(stdout);
   fflush(stderr);

   pid_t pid=fork();
   switch(pid)
   {
   case(0): // child
   {
      pid=getpid();
      detach();
      printf(_("[%u] Started.  %s\n"),(unsigned)pid,SMTask::now.IsoDateTime());
      SMTaskRef<AcceptTermFD> term_acceptor(new AcceptTermFD());
      for(;;)
      {
	 SMTask::Schedule();
	 if(Job::NumberOfJobs()==0)
	    break;
	 SMTask::Block();
	 if(term_acceptor->Accepted()) {
	    hook_signals();
	    top_exec->SetInteractive();
	    top_exec->SetStatusLine(new StatusLine(1));
	    top_exec->SetCmdFeeder(new ReadlineFeeder(0));
	    for(;;)
	    {
	       SMTask::Schedule();
	       if(top_exec->Done() || term_acceptor->Detached()) {
		  if(Job::NumberOfJobs()>0) {
		     printf(_("[%u] Detaching from the terminal to complete transfers...\n"),(unsigned)pid);
		  } else if(top_exec->Done()) {
		     printf(_("[%u] Exiting and detaching from the terminal.\n"),(unsigned)pid);
		  }
		  fflush(stdout);
		  term_acceptor->Detach();
		  detach();
		  printf(_("[%u] Detached from terminal. %s\n"),(unsigned)pid,SMTask::now.IsoDateTime());
		  break;
	       }
	       SMTask::Block();
	    }
	 }
      }
      top_exec->AtExitBg();
      top_exec->AtTerminate();
      top_exec->WaitDone();
      printf(_("[%u] Finished. %s\n"),(unsigned)pid,SMTask::now.IsoDateTime());
      break;
   }
   default: // parent
      printf(_("[%u] Moving to background to complete transfers...\n"),
	       (unsigned)pid);
      fflush(stdout);
      _exit(0);
   case(-1):
      perror("fork()");
   }
}

int lftp_slot(int count,int key)
{
   if(!top_exec)
      return 0;
   char slot[2];
   slot[0]=key;
   slot[1]=0;
   top_exec->ChangeSlot(slot);
   lftp_rl_set_prompt(top_exec->MakePrompt());
   return 0;
}

void  source_if_exist(const char *rc)
{
   if(access(rc,R_OK)!=-1)
   {
      top_exec->FeedCmd("source ");
      top_exec->FeedCmd(rc);
      top_exec->FeedCmd("\n");
   }
}

static void tty_clear()
{
   if(top_exec)
      top_exec->pre_stdout();
}

// look for the option, remove it and return true if found
static bool pick_option(int& argc,char **argv,const char *option)
{
   for(int i=1; i<argc; i++) {
      if(!strcmp(argv[i],option)) {
	 // remove the option, move trailing NULL too.
	 memmove(argv+i,argv+i+1,(argc-i)*sizeof(*argv));
	 argc--;
	 return true;
      }
      if(!strcmp(argv[i],"--"))	 // end of options
	 break;
   }
   return false;
}

char *program_name;

int   main(int argc,char **argv)
{
   program_name=argv[0];

#ifdef SOCKS4
   SOCKSinit(program_name);
#endif

   setlocale (LC_ALL, "");
   setlocale (LC_NUMERIC, "C");
   bindtextdomain (PACKAGE, LOCALEDIR);
   textdomain (PACKAGE);

   CmdExec::RegisterCommand("history",cmd_history,
	 N_("history -w file|-r file|-c|-l [cnt]"),
	 N_(" -w <file> Write history to file.\n"
	 " -r <file> Read history from file; appends to current history.\n"
	 " -c  Clear the history.\n"
	 " -l  List the history (default).\n"
	 "Optional argument cnt specifies the number of history lines to list,\n"
	 "or \"all\" to list all entries.\n"));
   CmdExec::RegisterCommand("attach",cmd_attach,"attach [PID]",
      N_("Attach the terminal to specified backgrounded lftp process.\n"));

   top_exec=new CmdExec(0,0);
   hook_signals();
   top_exec->SetStatusLine(new StatusLine(1));
   Log::global=new Log("debug");
   Log::global->SetCB(tty_clear);

   source_if_exist(SYSCONFDIR"/lftp.conf");

   if(!pick_option(argc,argv,"--norc")) {
      const char *home=getenv("HOME");
      if(home)
	 source_if_exist(dir_file(home,".lftprc"));
      home=get_lftp_config_dir();
      if(home)
	 source_if_exist(dir_file(home,"rc"));
   }

   top_exec->WaitDone();
   top_exec->SetTopLevel();
   top_exec->Fg();

   Ref<ArgV> args(new ArgV(argc,argv));
   args->setarg(0,"lftp");

   lftp_feeder=new ReadlineFeeder(args);

   top_exec->ExecParsed(args.borrow());
   top_exec->WaitDone();
   int exit_code=top_exec->ExitCode();

   top_exec->AtExit();
   top_exec->WaitDone();

   if(Job::NumberOfJobs()>0)
   {
      top_exec->SetInteractive(false);
      move_to_background();
   }
   else
   {
      top_exec->AtExitFg();
      top_exec->AtTerminate();
      top_exec->WaitDone();
   }
   top_exec->KillAll();
   top_exec=0;

   Job::Cleanup();
   ConnectionSlot::Cleanup();
   SessionPool::ClearAll();
   FileAccess::ClassCleanup();
   ProcWait::DeleteAll();
   IdNameCacheCleanup();
   SignalHook::Cleanup();
   Log::Cleanup();
   SMTask::Cleanup();
   return exit_code;
}
